rules/no-useless-promise-resolve-reject.js

```diff
@@ -0,0 +1,117 @@
+	[RETURN_REJECT]: 'Prefer `throw error` over `return Promise.reject(error)`.',
+	[YIELD_RESOLVE]: 'Prefer `yield value` over `yield Promise.resolve(value)`.',
+	[YIELD_REJECT]: 'Prefer `throw value` over `yield Promise.reject(value)`.',
+};
+
+const promiseResolveRejectSelector = methodCallSelector({
+	object: 'Promise',
+	methods: ['resolve', 'reject'],
+	allowComputed: true,
+});
+const asyncArrowFunctionReturnSelector = `ArrowFunctionExpression[async=true] > ${promiseResolveRejectSelector}`;
+const returnStatementSelector = `ReturnStatement > ${promiseResolveRejectSelector}`;
+const yieldExpressionSelector = `YieldExpression > ${promiseResolveRejectSelector}`;
+
+const functionTypes = new Set([
+	'ArrowFunctionExpression',
+	'FunctionDeclaration',
+	'FunctionExpression',
+]);
+const getParentFunction = node => {
+	let {parent} = node;
+	while (parent && !functionTypes.has(parent.type)) {
+		parent = parent.parent;
+	}
+
+	return parent;
+};
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const sourceCode = context.getSourceCode();
+
+	/** @param {(isResolve: boolean, valueString: string, value) => import('eslint').Rule.ReportFixer} createFix */
+	const createProblem = (node, createFix, resolveMessage = RETURN_RESOLVE, rejectMessage = RETURN_REJECT) => {
+		const isResolve = node.callee.property.name === 'resolve';
+		const value = node.arguments.length === 1 ? node.arguments[0] : undefined;
+		return {
+			node,
+			messageId: isResolve ? resolveMessage : rejectMessage,
+			fix: value === undefined ? undefined : createFix(isResolve, sourceCode.getText(value), value),
+		};
+	};
+
+	return {
+		[asyncArrowFunctionReturnSelector](node) {
+			return createProblem(
+				node,
+				(isResolve, value) => fixer =>
+					fixer.replaceText(
+						node,
+						isResolve
+							// Turns `=> Promise.resolve(value)` into `=> value`
+							? value
+							// Turns `=> value` into `=> { throw value }`
+							: `{ throw ${value}; }`,
+					),
+			);
+		},
+		[returnStatementSelector](node) {
+			const parentFunction = getParentFunction(node);
+			if (!parentFunction || !parentFunction.async) {
+				return;
+			}
+
+			return createProblem(
+				node,
+				(isResolve, value) => fixer => isResolve
+					// Turns `return Promise.resolve(value)` into `return value`
+					? fixer.replaceText(node, value)
+					// Turns `return Promise.reject(value)` into `throw value`
+					: fixer.replaceText(node.parent, `throw ${value};`),
+			);
+		},
+		[yieldExpressionSelector](node) {
+			const parentFunction = getParentFunction(node);
+			if (!parentFunction || !parentFunction.async || !parentFunction.generator) {
+				return;
+			}
+
+			return createProblem(
+				node,
+				(isResolve, valueString, value) => fixer => isResolve
+					// Turns `yield Promise.resolve(value)` into `yield value`
+					? fixer.replaceText(node, value.type === 'SequenceExpression' ? `(${valueString})` : valueString)
```

Comment:
If https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1623/files#r771216741 is addressed, this `SequenceExpression` should not needed, in most case, because it's originally a call argument, we just keep the parens, won't need insert. But I'm not sure if there are edge cases don't need parens in argument, but neeeded in `yield`.

Reply:
Check `should-add-parentheses-to-*.js` in `rules/utils` directory, need implment similar logic.